websockets/echo/src/server.rs

```diff
@@ -11,68 +10,84 @@ const CLIENT_TIMEOUT: Duration = Duration::from_secs(10);
-        ctx.run_interval(HEARTBEAT_INTERVAL, |act, ctx| {
-            // check client heartbeats
-            if Instant::now().duration_since(act.hb) > CLIENT_TIMEOUT {
-                // heartbeat timed out
-                println!("Websocket Client heartbeat failed, disconnecting!");
-
-                // stop actor
-                ctx.stop();
-
-                // don't try to send a ping
-                return;
-            }
+    async fn pre_start(
+        &self,
+        myself: ActorRef<Self::Msg>,
+        session: Self::Arguments,
+    ) -> Result<Self::State, ActorProcessingErr> {
+        myself.send_interval(HEARTBEAT_INTERVAL, || AsMessage::Hb);
 
-            ctx.ping(b"");
-        });
+        Ok((Instant::now(), Some(session)))
     }
-}
 
-impl Actor for MyWebSocket {
-    type Context = ws::WebsocketContext<Self>;
+    async fn handle(
+        &self,
+        myself: ActorRef<Self::Msg>,
+        message: Self::Msg,
+        state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        match message {
+            AsMessage::Hb => {
+                // check client heartbeats
+                if Instant::now().duration_since(state.0) > CLIENT_TIMEOUT {
+                    // heartbeat timed out
+                    println!("Websocket Client heartbeat failed, disconnecting!");
 
-    /// Method is called on actor start. We start the heartbeat process here.
-    fn started(&mut self, ctx: &mut Self::Context) {
-        self.hb(ctx);
-    }
-}
+                    let _ = state.1.take().unwrap().close(None).await;
 
-/// Handler for `ws::Message`
-impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for MyWebSocket {
-    fn handle(&mut self, msg: Result<ws::Message, ws::ProtocolError>, ctx: &mut Self::Context) {
-        // process websocket messages
-        println!("WS: {msg:?}");
-        match msg {
-            Ok(ws::Message::Ping(msg)) => {
-                self.hb = Instant::now();
-                ctx.pong(&msg);
-            }
-            Ok(ws::Message::Pong(_)) => {
-                self.hb = Instant::now();
+                    // stop actor
+                    myself.stop(None);
+
+                    // don't try to send a ping
+                } else {
+                    state.1.as_mut().unwrap().ping(b"").await.unwrap();
+                }
             }
-            Ok(ws::Message::Text(text)) => ctx.text(text),
-            Ok(ws::Message::Binary(bin)) => ctx.binary(bin),
-            Ok(ws::Message::Close(reason)) => {
-                ctx.close(reason);
-                ctx.stop();
+
+            AsMessage::Ws(msg) => {
+                // process websocket messages
+                println!("WS: {msg:?}");
+
+                match msg {
+                    AggregatedMessage::Ping(msg) => {
+                        state.0 = Instant::now();
+                        state.1.as_mut().unwrap().pong(&msg).await.unwrap();
```

Comment:
if killing the actor is intended with the `unwrap()`, you also can just use `?` notation as we capture anything that can be typed to an `Error`. The benefit of that is you maintain the error source in the Error trace.

Reply:
Ahh right, hadn't spotted that ActorProcessingErr was an alias.